ansible/deploy.yml

```diff
@@ -115,6 +115,9 @@
         state: restarted
       when: hostvars['localhost']['database_node_ip'] != node_ip
 
+    - name: Load distro specific vars
```

Comment:
Should we have a more explicit check earlier that you're using a supported distro and error out with a descriptive warning if you're not?

Reply:
Yeah, good question. I guess I was trying to find the balance balance between current stance of "only Ubuntu is supported" and making it slightly more flexible. Are you wanting this to also work for wider Debian family (noting a play is `debian.yml` rather than `ubuntu.yml` or even something generic)? It's easy to make it also work on Debian by adding a new file or linking to the Ubuntu one. We can also make it distro "family" based and put more logic into the vars file if you prefer that approach (depends how many other non-Ubuntu Debian family distros you wanna support).

As you point out, the issue with the approach in this patch is that if someone runs it on anything other than Ubuntu it will fail, so depends what you're after. If you want it to have same net effect as current code then we skip distro specific vars and have Ubuntu flavoured set of defaults. This way we don't load distro specific vars yet (no failure on non-Ubuntu distros) which would have same effect as the existing hard-coded list) and continue to assume people are running it on something that works. This is one of the reasons I've made it a variable, so that if it's on an unsupported distro, people can still override that var and it'll work (well, so long as I remove the hard failure when loading distro specific vars).

If you want to error on unsupported distros, then I'd probably do a set of checks at the beginning and bail earlier. But doing so might deter people from making it work on other distros for you, if you're interested in that in the future.

So basically, happy to re-work it which ever way you want to go...

Reply:
Yeah, so there is some definite history here. The manifesto for Shaken Fist is something like "do something only one way, and that way should be the simplest one possible". So at the moment I have zero interest in supporting 87 Linux distros, or FreeBSD or whatever. That said, if someone could argue that supporting other distros wasn't actually more complicated, I'd at least politely listen.

The debian.yml thing is because back in the day the only distro I could find docs for virt-in-virt on GCP for was Debian, so that's what I used. In the end I got it working with Ubuntu and moved back to that because the debian package names were weirdly different and annoying. The ansible filename just kind of hung around.

That said, deploy.yml does have guard rails. I see checking if you're using Ubuntu as a reasonable install safety check, which positions us so we could do other distros later if we thought of a good reason to. Its similar in my mind to how we run kvm-ok to make sure that KVM will actually work on your hypervisors.

I like the idea of doing checks earlier, not making it deliberately hard to support other distros (if someone wants to maintain a private branch for their distro of choice I would definitely not stop that), and exitting with a message saying if someone wants to make a business case for their distro being supported we're willing to talk. Does that make sense?

Reply:
Yeah, that makes sense. I was aware of the history which is why I went down the Ubuntu only path, but with at least some added flexibility/groundwork if others wanted to support other distros without destroying the code. So I think you're saying, let's only support Ubuntu but not stop others from trying to run it if they know what they're doing. I'll work on an update and push it for comment.

Reply:
Yes, sounds good. Thanks.